sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UnusedPrivateMember.cs

```diff
@@ -80,486 +65,209 @@ protected sealed override void Initialize(SonarAnalysisContext context)
 
-                            var propertyAccessorAccess = new Dictionary<IPropertySymbol, AccessorAccess>();
+                            VisitDeclaringReferences(namedType, usageCollector);
 
-                            CollectUsedSymbols(declarationCollector, usedSymbols, declaredPrivateSymbols, propertyAccessorAccess);
-                            CollectUsedSymbolsFromCtorInitializerAndCollectEmptyCtors(declarationCollector,
-                                usedSymbols, emptyConstructors);
-
-                            ReportIssues(cc, usedSymbols, declaredPrivateSymbols, emptyConstructors, fieldLikeSymbols);
-                            ReportUnusedPropertyAccessors(cc, usedSymbols, declaredPrivateSymbols, propertyAccessorAccess);
+                            var diagnostics = GetDiagnostics(usageCollector, removableSymbolsCollector.PrivateSymbols, "private",
+                                removableSymbolsCollector.FieldLikeSymbols);
+                            foreach (var diagnostic in diagnostics)
+                            {
+                                cc.ReportDiagnosticIfNonGenerated(diagnostic, cc.Compilation);
+                            }
                         },
                         SymbolKind.NamedType);
 
                     c.RegisterCompilationEndAction(
-                       cc =>
-                       {
-                           if (!shouldRaise)
-                           {
-                               return;
-                           }
-
-                           foreach (var diagnostic in possibleUnusedInternalMembers)
-                           {
-                               cc.ReportDiagnosticIfNonGenerated(diagnostic, cc.Compilation);
-                           }
-                       });
-                });
-        }
+                        cc =>
+                        {
+                            var foundInternalsVisibleTo = cc.Compilation.Assembly
+                                .GetAttributes(KnownType.System_Runtime_CompilerServices_InternalsVisibleToAttribute)
+                                .Any();
 
-        private static bool IsInternalVisibleToAttribute(AttributeSyntax attribute, SemanticModel semanticModel)
-        {
+                            if (foundInternalsVisibleTo)
+                            {
+                                return;
+                            }
+
+                            var usageCollector = new SymbolUsageCollector(
+                                c.Compilation.GetSemanticModel,
+                                removableInternalMembers.Select(s => s.Name).ToHashSet());
 
-            return semanticModel.GetSymbolInfo(attribute).Symbol is IMethodSymbol attributeConstructor &&
-                attributeConstructor.ContainingType.Is(KnownType.System_Runtime_CompilerServices_InternalsVisibleToAttribute);
+                            foreach (var symbol in allNamedTypes)

```

Comment:
Corner case: walking named types won't catch references to fields/properties/methods that are in assembly-level attributes e.g. _AssemblyVersion_. The only way to catch those would be to walk all of the syntax trees in the assembly. I can't think of any other ways of refering to a field/property etc outside of type, but there might be some.

I suspect we will get some false-positives from this, but not many so there's probably no need to handle this case now. However, it would be adding a test case and marking it as a known false positive.

'''
[assembly: AssemblyCompany(Ns1.MyAppDetails.AppName)]
namespace Ns1
{
    internal static class MyAppDetails
    {
        public const string AppName = "myApp";
    }
}
'''